src/components/UpgradeButton.js

```diff
@@ -1,22 +1,63 @@
-import React from "react";
+import React, { useState } from "react";
 import PropTypes from "prop-types";
 
-const UpgradeButton = ({ type }) => {
+import Button from "@material-ui/core/Button";
+import Menu from "@material-ui/core/Menu";
+import MenuItem from "@material-ui/core/MenuItem";
+
+const UpgradeButton = ({
+  type,
+  upgrades,
+  upgradeFunction,
+  unitKey,
+  upgradeKey
+}) => {
+  const [anchorEl, setAnchorEl] = useState(null);
+  const [selectedUpgrade, setSelectedUpgrade] = useState(undefined);
+
+  const handleClick = event => {
+    setAnchorEl(event.currentTarget);
+  };
+
+  const handleClose = upgrade => {
+    setSelectedUpgrade(upgrade);
+    upgradeFunction(unitKey, upgradeKey, upgrade);
+    setAnchorEl(null);
+  };
+
   return (
-    <button
-      onClick={e => {
-        // e.preventDefault();
-        e.stopPropagation();
-        alert(`todo: Add a ${type} upgrade`);
-      }}
-    >
-      {type}
-    </button>
+    <>
+      <Button variant="contained" onClick={handleClick}>
+        {selectedUpgrade ? selectedUpgrade.name : type}
+      </Button>
+      <Menu
+        id="upgrade-menu"
+        anchorEl={anchorEl}
+        keepMounted
+        open={Boolean(anchorEl)}
+        onClose={() => handleClose(selectedUpgrade)}
+      >
+        {upgrades &&
```

Comment:
Small suggestion:

```suggestion
        {(upgrades && upgrades.length > 0) &&
            upgrades.map(upgrade => ...
```

The first two in this chain are conditionals and the third is the outcome. IMO helps with clarity to explicitly state the conditions on one line with the outcome on the next. If the conditional line becomes too long, good time to refactor into a function.